pkg/openid4ci/openid4ci.go

```diff
@@ -73,17 +76,181 @@ func (i *Interaction) Authorize() (*AuthorizeResult, error) {
+
+	// TODO: Implement trusted list type of mechanism? The gosec warning (correctly) warns about using a variable URL.
+	response, err := http.Get(metadataEndpoint) //nolint: noctx,gosec // TODO: To be re-evaluated later
+	if err != nil {
+		return nil, err
+	}
+
+	responseBytes, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received status code [%d] with body [%s] from issuer's "+
+			"OpenID configuration endpoint", response.StatusCode, string(responseBytes))
+	}
+
+	defer func() {
+		errClose := response.Body.Close()
+		if errClose != nil {
+			println(fmt.Sprintf("failed to close response body: %s", errClose.Error()))
+		}
+	}()
+
+	var metadata issuerMetadata
+
+	err = json.Unmarshal(responseBytes, &metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response from the issuer's "+
+			"OpenID configuration endpoint: %w", err)
+	}
+
+	return &metadata, err
+}
+
+func (i *Interaction) buildTokenURLWithQueryParameters(userPIN, tokenEndpoint string) (*url.URL, error) {
+	tokenEndpointURL, err := url.Parse(tokenEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	query := tokenEndpointURL.Query()
+
+	query.Add("grant_type", "urn:ietf:params:oauth:grant-type:pre-authorized_code")
+	query.Add("pre-authorized_code", i.initiationRequest.PreAuthorizedCode)
+	query.Add("user_pin", userPIN)
+
+	tokenEndpointURL.RawQuery = query.Encode()
+
+	return tokenEndpointURL, nil
+}
+
+func (i *Interaction) getTokenResponse(tokenEndpointURL *url.URL) (*tokenResponse, error) {
+	response, err := http.Post(tokenEndpointURL.String(), //nolint: noctx // TODO: To be re-evaluated later
+		"application/x-www-form-urlencoded", http.NoBody)
+	if err != nil {
+		return nil, err
+	}
+
+	responseBytes, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received status code [%d] with body [%s] from issuer's token endpoint",
+			response.StatusCode, string(responseBytes))
 	}
 
-	return []*CredentialResponse{getSampleCredentialResponse()}, nil
+	defer func() {
+		errClose := response.Body.Close()
+		if errClose != nil {
+			println(fmt.Sprintf("failed to close response body: %s", errClose.Error()))
+		}
+	}()
+
+	var tokenResp tokenResponse
+
+	err = json.Unmarshal(responseBytes, &tokenResp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response from the issuer's token endpoint: %w", err)
+	}
+
+	return &tokenResp, nil
 }
 
-func getSampleCredentialResponse() *CredentialResponse {
-	return &CredentialResponse{
-		Credential: &verifiable.Credential{ID: "SampleID"},
-		Format:     "SampleFormat",
+func (i *Interaction) getCredentialResponse(credentialType, credentialEndpoint,
+	accessToken string,
+) (*CredentialResponse, error) {
+	credentialReq := &credentialRequest{
+		Type: credentialType,
+		Proof: proof{
+			ProofType: "jwt",
```

Comment:
do not hardcode to jwt, add a TODO if we are hard wiring it for now.
it should be able to be overridden by options.

Reply:
Added TODO.